src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -637,9 +637,23 @@ Pipes MergeTreeDataSelectExecutor::readFromParts(
             reader_settings,
             result_projection);
     }
-    else if ((settings.optimize_read_in_order || settings.optimize_aggregation_in_order) && query_info.input_order_info)
+    else if (((settings.optimize_read_in_order || settings.optimize_aggregation_in_order) && query_info.input_order_info) ||
+            (settings.optimize_distinct_in_order && query_info.distinct_order_info))
     {
-        size_t prefix_size = query_info.input_order_info->order_key_prefix_descr.size();
+        bool prefer_distinct_order = false;
+        size_t prefix_size = 0, distinct_prefix_size = 0;
+        if (query_info.input_order_info)
+            prefix_size = query_info.input_order_info->order_key_prefix_descr.size();
+        else
+            prefer_distinct_order = true;
+        if (query_info.distinct_order_info)
+            distinct_prefix_size = query_info.distinct_order_info->order_key_prefix_descr.size();
+        if (!settings.optimize_aggregation_in_order && distinct_prefix_size > prefix_size)
+        {
+            prefer_distinct_order = true;
+            prefix_size = distinct_prefix_size;
+        }
+        std::cerr << std::boolalpha << prefer_distinct_order << "\n";
```

Comment:
Maybe it's better to do it at previous level (`InterpreterSelectQuery` or `SyntaxAnalyzer`)?